File: app/src/main/java/com/example/myapplication/JobDispatchers/CommentDispatch.java
package com.example.myapplication.JobDispatchers;

import android.content.Context;
import androidx.annotation.NonNull;
import androidx.work.Worker;
import androidx.work.WorkerParameters;
import com.example.myapplication.API.ApiUtils.ApiClient;
import com.example.myapplication.API.DAO.CommentDao;
import com.example.myapplication.Models.CommentModel;
import com.example.myapplication.Models.CommentResponseBody;
import com.example.myapplication.Room.PersistenceDatabase;
import retrofit2.Call;
import retrofit2.Response;

import java.io.IOException;
import java.util.List;
import java.util.concurrent.*;

public class CommentDispatch extends Worker {

    private static CommentDao commentDao;
    private static com.example.myapplication.Room.Dao.CommentDao persist_comment_dao;
    private ApiClient apiClient;
    private PersistenceDatabase persistenceDatabase;

    public CommentDispatch(@NonNull Context context, @NonNull WorkerParameters workerParams) {
        super(context, workerParams);

        persistenceDatabase = PersistenceDatabase.getInstance(context);
        persist_comment_dao = persistenceDatabase.commentDao();
        apiClient = new ApiClient(context);
        commentDao = apiClient.getClient(CommentDao.class);
    }

    @NonNull
    @Override
    public Result doWork() {
        boolean status = false;
        try {
            status = sendComments();
        } catch (ExecutionException e) {
            e.printStackTrace();
        } catch (InterruptedException e) {
            e.printStackTrace();
        }

        if(!status){
            return Result.retry();
        }
        return Result.success();
    }

    private boolean sendComments() throws ExecutionException, InterruptedException {
        ExecutorService executorService = Executors.newSingleThreadExecutor();
        Future<Boolean> future = executorService.submit(new CommentRunnable());

        while (!future.isDone()){
            try {
                Thread.sleep(1000);
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }
        return future.get();
    }

    private static class CommentRunnable implements Callable<Boolean> {

        @Override
        public Boolean call() {
            boolean value = false;
            List<CommentModel> comments = persist_comment_dao.getUnsentComments();
            for (CommentModel commentModel : comments) {
                Call<CommentResponseBody> call = commentDao.sendComment(commentModel);
                try {
                    Response<CommentResponseBody> response = call.execute();
                    if (response.isSuccessful()) {
                        assert response.body() != null;
                        persist_comment_dao.insertCommentAndUser(response.body());
                    }
                    value = true;
                } catch (IOException e) {
                    e.printStackTrace();
                    value = false;
                }
            }
            return value;
        }
    }
}

File: app/src/main/java/com/example/myapplication/JobDispatchers/CommentDispatch.kt
package com.example.myapplication.JobDispatchers

import android.content.Context
import android.util.Log
import androidx.work.Worker
import androidx.work.WorkerParameters
import com.example.myapplication.API.ApiUtils.ApiClient
import com.example.myapplication.API.DAO.CommentDao
import com.example.myapplication.Room.CommentDatabase
import com.example.myapplication.Room.PersistenceDatabase
import com.example.myapplication.repository.CommentRepository
import kotlinx.coroutines.*
import retrofit2.HttpException
import java.io.IOException

class CommentDispatch(context: Context, workerParameters: WorkerParameters) : Worker(context, workerParameters) {

    private val apiClient = ApiClient.getInstance(context).getClient(CommentDao::class.java)
    private val commentRepository: CommentRepository = CommentRepository(context)

    override fun doWork() : Result {
        var status = false

            val job = CoroutineScope(Dispatchers.IO).launch{
                status = withContext(Dispatchers.IO) { sendComments() }
            }

        while(!job.isCompleted){
            try{
                CoroutineScope(Dispatchers.IO).launch {
                    delay(1500)
                }
            }catch (exception: InterruptedException){
                exception.printStackTrace()
            }
        }


        if(!status){
            return Result.retry()
        }
        return Result.success()
    }

    private suspend fun sendComments(): Boolean = withContext(Dispatchers.IO) {
        var status = false
        val comments = commentRepository.getUnsentComments()
        if(comments.isNotEmpty()){
            for ( commentModel in comments) {
                status = try {
                    val response = apiClient.sendComment(commentModel)
                    commentRepository.updateCommentInDB(response)
                    true
                } catch (e : HttpException) {
                    false
                }
            }
            status
        }else{
            true
        }
    }
}
